pkg/controller/dnsendpoint/service.go

```diff
@@ -58,56 +58,52 @@ func getServiceDNSEndpoints(obj interface{}) ([]*feddnsv1a1.Endpoint, error) {
-
-		dnsNames := []string{
-			strings.Join([]string{commonPrefix, zone, region, dnsObject.Status.Domain}, "."), // zone level
-			strings.Join([]string{commonPrefix, region, dnsObject.Status.Domain}, "."),       // region level, one up from zone level
-			strings.Join([]string{commonPrefix, dnsObject.Status.Domain}, "."),               // global level, one up from region level
-			"", // nowhere to go up from global level
-		}
+	ttl := dnsObject.Spec.RecordTTL
+	if ttl == 0 {
+		ttl = defaultDNSTTL
+	}
 
-		var zoneTargets, regionTargets, globalTargets feddnsv1a1.Targets
-		for _, clusterDNS := range dnsObject.Status.DNS {
-			if clusterDNS.Zone == zone {
-				zoneTargets = append(zoneTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		var zoneDNSName string
+		regionDNSName := strings.Join([]string{commonPrefix, clusterDNS.Region, dnsObject.Status.Domain}, ".") // region level, one up from zone level
+		globalDNSName := strings.Join([]string{commonPrefix, dnsObject.Status.Domain}, ".")                    // global level, one up from region level
+
+		// Zone endpoints
+		for _, zone := range clusterDNS.Zones {
+			zoneDNSName = strings.Join([]string{commonPrefix, zone, clusterDNS.Region, dnsObject.Status.Domain}, ".")
+			zoneTargets := ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)
+			zoneEndpoint, err := generateEndpointForServiceDNSObject(zoneDNSName, zoneTargets, regionDNSName, ttl, labels)
+			if err != nil {
+				return nil, err
 			}
+			endpoints = append(endpoints, zoneEndpoint)
 		}
 
-		for _, clusterDNS := range dnsObject.Status.DNS {
-			if clusterDNS.Region == region {
-				regionTargets = append(regionTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
-			}
+		// Region endpoints
+		regionTargets := ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)
+		regionEndpoint, err := generateEndpointForServiceDNSObject(regionDNSName, regionTargets, globalDNSName, ttl, labels)
+		if err != nil {
+			return nil, err
 		}
+		endpoints = append(endpoints, regionEndpoint)
 
-		for _, clusterDNS := range dnsObject.Status.DNS {
-			globalTargets = append(globalTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		// Global endpoints
+		globalTargets := ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)
+		globalEndpoint, err := generateEndpointForServiceDNSObject(globalDNSName, globalTargets, "", ttl, labels)
+		if err != nil {
+			return nil, err
 		}
+		endpoints = append(endpoints, globalEndpoint)
+	}
 
-		targets := [][]string{zoneTargets, regionTargets, globalTargets}
-
-		ttl := dnsObject.Spec.RecordTTL
-		if ttl == 0 {
-			ttl = defaultDNSTTL
-		}
-		for i, target := range targets {
-			endpoint, err := generateEndpointForServiceDNSObject(dnsNames[i], target, dnsNames[i+1], ttl, labels)
-			if err != nil {
-				return nil, err
-			}
-			endpoints = append(endpoints, endpoint)
-		}
-		if dnsObject.Spec.DNSPrefix != "" {
```

Comment:
(No action required) Oops!  I guess this wasn't a problem because of the dedup?

Reply:
yes, this was masked before because of dedup.